libs/libcommon/src/libcommon/orchestrator.py

```diff
@@ -728,6 +801,153 @@ def is_big(config: str) -> bool:
             self.add_task(CreateJobsTask(job_infos=job_infos_to_create))
 
 
+class SmartUpdateImpossibleBecauseCacheIsEmpty(Exception):
+    pass
+
+
+class SmartUpdateImpossibleBecauseOfUpdatedFiles(Exception):
+    pass
+
+
+class SmartUpdateImpossibleBecauseOfUpdatedYAMLField(Exception):
+    pass
+
+
+class SmartUpdateImpossibleBecauseCachedRevisionIsNotParentOfNewRevision(Exception):
+    pass
+
+
+@dataclass
+class SmartDatasetUpdatePlan(Plan):
+    dataset: str
+    revision: str
+    hf_endpoint: str
+    old_revision: str
+    processing_graph: ProcessingGraph = field(default=processing_graph)
+    storage_clients: Optional[list[StorageClient]] = None
+    hf_token: Optional[str] = None
+
+    cached_revision: str = field(init=False)
+    diff: str = field(init=False)
+    files_impacted_by_commit: set[str] = field(init=False)
+    updated_yaml_fields_in_dataset_card: list[str] = field(init=False)
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+        cache_kinds = [
+            processing_step.cache_kind for processing_step in self.processing_graph.get_first_processing_steps()
+        ]
+        # Try to be robust to a burst of webhooks or out-of-order webhooks
+        # by waiting up to 2 seconds for a coherent state
+        for retry in range(3):
+            cache_entries_df = get_cache_entries_df(
+                dataset=self.dataset,
+                cache_kinds=cache_kinds,
+            )
+            if len(cache_entries_df) == 0:
+                raise SmartUpdateImpossibleBecauseCacheIsEmpty(f"Failed to smart update to {self.revision[:7]}")
+            self.cached_revision = cache_entries_df.sort_values("updated_at").iloc[-1]["dataset_git_revision"]
```

Comment:
shouldn't we check that all the entries have the same value?

Reply:
yup good idea, I'm adding an error in that case